kernel/arch/dreamcast/hardware/pvr/pvr_prim.c

```diff
@@ -338,385 +268,186 @@ void pvr_sprite_cxt_txr(pvr_sprite_cxt_t *dst, pvr_list_t list,
+        | FIELD_PREP(PVR_TA_PM2_DSTENABLE, src->blend.dst_enable2)
+        | FIELD_PREP(PVR_TA_PM2_FOG, src->gen.fog_type2)
+        | FIELD_PREP(PVR_TA_PM2_CLAMP, src->gen.color_clamp2)
+        | FIELD_PREP(PVR_TA_PM2_ALPHA, src->gen.alpha2);
 
     if(src->txr2.enable == PVR_TEXTURE_DISABLE) {
-        dst->mode2_1 = mode2[1];
-        dst->mode3_1 = 0;
+        mode3 = 0;
     }
     else {
-        mode2[1] |= (src->txr2.alpha << PVR_TA_PM2_TXRALPHA_SHIFT) & PVR_TA_PM2_TXRALPHA_MASK;
-        mode2[1] |= (src->txr2.uv_flip << PVR_TA_PM2_UVFLIP_SHIFT) & PVR_TA_PM2_UVFLIP_MASK;
-        mode2[1] |= (src->txr2.uv_clamp << PVR_TA_PM2_UVCLAMP_SHIFT) & PVR_TA_PM2_UVCLAMP_MASK;
-        mode2[1] |= (src->txr2.filter << PVR_TA_PM2_FILTER_SHIFT) & PVR_TA_PM2_FILTER_MASK;
-        mode2[1] |= (src->txr2.mipmap_bias << PVR_TA_PM2_MIPBIAS_SHIFT) & PVR_TA_PM2_MIPBIAS_MASK;
-        mode2[1] |= (src->txr2.env << PVR_TA_PM2_TXRENV_SHIFT) & PVR_TA_PM2_TXRENV_MASK;
-
-        switch(src->txr2.width) {
-            case 8:
-                u = 0;
-                break;
-            case 16:
-                u = 1;
-                break;
-            case 32:
-                u = 2;
-                break;
-            case 64:
-                u = 3;
-                break;
-            case 128:
-                u = 4;
-                break;
-            case 256:
-                u = 5;
-                break;
-            case 512:
-                u = 6;
-                break;
-            case 1024:
-                u = 7;
-                break;
-            default:
-                assert_msg(0, "Invalid texture U size");
-                u = 0;
-                break;
-        }
-
-        switch(src->txr2.height) {
-            case 8:
-                v = 0;
-                break;
-            case 16:
-                v = 1;
-                break;
-            case 32:
-                v = 2;
-                break;
-            case 64:
-                v = 3;
-                break;
-            case 128:
-                v = 4;
-                break;
-            case 256:
-                v = 5;
-                break;
-            case 512:
-                v = 6;
-                break;
-            case 1024:
-                v = 7;
-                break;
-            default:
-                assert_msg(0, "Invalid texture V size");
-                v = 0;
-                break;
-        }
-
-        mode2[1] |= (u << PVR_TA_PM2_USIZE_SHIFT) & PVR_TA_PM2_USIZE_MASK;
-        mode2[1] |= (v << PVR_TA_PM2_VSIZE_SHIFT) & PVR_TA_PM2_VSIZE_MASK;
-
-        dst->mode2_1 = mode2[1];
-
-        /* Polygon mode 3 (within volume) */
-        mode3[1]  = (src->txr2.mipmap << PVR_TA_PM3_MIPMAP_SHIFT) & PVR_TA_PM3_MIPMAP_MASK;
-        mode3[1] |= (src->txr2.format << PVR_TA_PM3_TXRFMT_SHIFT) & PVR_TA_PM3_TXRFMT_MASK;
+        assert_msg(__builtin_popcount(src->txr2.width) == 1
+		   && src->txr2.width <= 1024, "Invalid texture U size");
```

Comment:
Just putting it out here that this is really clever (along with the ctz trick below), but it definitely threw me for a minute and made me have to look up what those builtins do... It might be a good idea to eventually add a comment for posterity explaining what in the world is going on there so that others don't get confused if they try to grok the code, but that's for another day...

Ultimately, most people aren't all that familiar with the weirder named gcc builtins.

Reply:
We can probably wrap those with macro names e.g. `count_bits()` and `count_trailing_zeros()`.